File: ios/Shared/Lab/ShowcaseModels.swift
import Foundation

// Codable mirrors of the engine's BenchmarkPayload JSON (camelCase keys), decoded with JSONDecode.

struct BenchmarkPoint: Codable, Sendable {
    let x: Double
    let y: Double
}

struct BenchmarkSeries: Codable, Sendable, Identifiable {
    let name: String
    let points: [BenchmarkPoint]
    var id: String { name }
}

struct SummaryStat: Codable, Sendable, Identifiable {
    let label: String
    let value: String
    var id: String { label }
}

struct BenchmarkPayload: Codable, Sendable {
    let kind: String
    let title: String
    let series: [BenchmarkSeries]
    let summary: [SummaryStat]
}

File: core/DotnetNativeInterop.Engine/Showcase/ShowcaseModels.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace DotnetNativeInterop.Engine;

/// <summary>One (x, y) sample in a benchmark series.</summary>
public sealed record BenchmarkPoint(double X, double Y);

/// <summary>A named series of benchmark points (e.g. "scalar" vs "SIMD").</summary>
public sealed record BenchmarkSeries(string Name, BenchmarkPoint[] Points);

/// <summary>A headline number for a benchmark (e.g. peak GFLOPS, speedup).</summary>
public sealed record SummaryStat(string Label, string Value);

/// <summary>
/// The structured result of a heavy-compute benchmark, embedded as JSON in
/// <see cref="FeatureRun.Result"/> so the native UI can chart it with Swift Charts.
/// </summary>
public sealed record BenchmarkPayload(string Kind, string Title, BenchmarkSeries[] Series, SummaryStat[] Summary);

/// <summary>
/// Source-generated JSON metadata for <see cref="BenchmarkPayload"/> (AOT-safe, no reflection). Also
/// registers <see cref="FeatureDescriptor"/>/its list shape so <see cref="RealPayloadBenchmark"/> can
/// serialize the language-feature catalog without a reflection fallback.
/// </summary>
[JsonSourceGenerationOptions(PropertyNamingPolicy = JsonKnownNamingPolicy.CamelCase)]
[JsonSerializable(typeof(BenchmarkPayload))]
[JsonSerializable(typeof(IReadOnlyList<FeatureDescriptor>))]
[JsonSerializable(typeof(FeatureDescriptor))]
internal sealed partial class ShowcaseJsonContext : JsonSerializerContext;

/// <summary>Serializes a <see cref="BenchmarkPayload"/> to camelCase JSON via the source-gen context.</summary>
public static class ShowcaseJson
{
    public static string Serialize(BenchmarkPayload payload) =>
        JsonSerializer.Serialize(payload, ShowcaseJsonContext.Default.BenchmarkPayload);
}
